### ISSUE TITLE
util-linux-2.39-rc3  error: redeclaration of 'enum fsconfig_command'

### ISSUE THREAD

**Issue by kladitt** [util-linux/util-linux#2186]
With util-linux-2.39-rc3 i get this error:
In file included from ./include/mount-api-utils.h:11,
                 from libmount/src/mountP.h:32,
                 from libmount/src/cache.c:34:
/usr/include/linux/mount.h:95:6: error: redeclaration of 'enum fsconfig_command'
   95 | enum fsconfig_command {
      |      ^~~~~~~~~~~~~~~~
In file included from libmount/src/libmount.h:40,
                 from libmount/src/mountP.h:31,
                 from libmount/src/cache.c:34:
/usr/lib/gcc/x86_64-unknown-linux-gnu/11.3.1/include-fixed/sys/mount.h:249:6: note: originally defined here

util-linux-2.38.1 compiles without any error

kernel 6.1.24, gcc 11.3.1 20230414, glibc 2.37.9000

**Comment by karelzak** [comment]
I did a small includes cleanup in the code (commit https://github.com/util-linux/util-linux/commit/7511ffb9d039a1d2d174e6915eff79b04690d6ac), but I'm not sure if it helps for your case. 

It seems like a conflict between /usr/include/linux/mount.h and /usr/include/sys/mount.h. Unfortunately, we need both files on many systems ...

**Comment by brauner** [comment]
Yeah, so this sounds like the `linux/mount.h` vs `sys/mount.h` mess where `linux/mount.h` has `defines` and `sys/mount.h` has enums. I have a pretty brutal hack in one of our codebases:

```
## This is a hack to prevent any inclusion ofr linux/mount.h which causes
## conflicts with sys/mount.h all over the place
srcconf.set('_LINUX_MOUNT_H', true)
```

and then I have a `mount_utils.h` that defines everything unless it's provided by `sys/mount.h`.

**Comment by kladitt** [comment]
On 20.04.23 12:59, Karel Zak wrote:
> I did a small includes cleanup in the code (commit https://github.com/util-linux/util-linux/commit/7511ffb9d039a1d2d174e6915eff79b04690d6ac), but I'm not sure if it helps for your case.
> 
> It seems like a conflict between /usr/include/linux/mount.h and /usr/include/sys/mount.h. Unfortunately, we need both files on many systems ...
> 

Karel, unfortunately your cleanup it did not help.

Other programms also suffer from this conflict.

-- 
Regards Klaus

**Comment by brauner** [comment]
Fwiw, it is fixed with:

```
commit 774058d72942249f71d74e7f2b639f77184160a6
Author:     Adhemerval Zanella <adhemerval.zanella@linaro.org>
AuthorDate: Wed Aug 10 14:24:47 2022 -0300
Commit:     Adhemerval Zanella <adhemerval.zanella@linaro.org>
CommitDate: Fri Aug 12 09:15:28 2022 -0300

    linux: Fix sys/mount.h usage with kernel headers

    Now that kernel exports linux/mount.h and includes it on linux/fs.h,
    its definitions might clash with glibc exports sys/mount.h.  To avoid
    the need to rearrange the Linux header to be always after glibc one,
    the glibc sys/mount.h is changed to:

      1. Undefine the macros also used as enum constants.  This covers prior
         inclusion of <linux/mount.h> (for instance MS_RDONLY).

      2. Include <linux/mount.h> based on the usual __has_include check
         (needs to use __has_include ("linux/mount.h") to paper over GCC
         bugs.

      3. Define enum fsconfig_command only if FSOPEN_CLOEXEC is not defined.
         (FSOPEN_CLOEXEC should be a very close proxy.)

      4. Define struct mount_attr if MOUNT_ATTR_SIZE_VER0 is not defined.
         (Added in the same commit on the Linux side.)

    This patch also adds some tests to check if including linux/fs.h and
    linux/mount.h after and before sys/mount.h does work.

    Checked on x86_64-linux-gnu.

    Reviewed-by: Florian Weimer <fweimer@redhat.com>
```

starting with glibc 2.37...

**Comment by kladitt** [comment]
On 20.04.23 16:33, Christian Brauner wrote:
> Fwiw, it is fixed with:
> 
> ```
> commit 774058d72942249f71d74e7f2b639f77184160a6
> Author:     Adhemerval Zanella ***@***.***>
> AuthorDate: Wed Aug 10 14:24:47 2022 -0300
> Commit:     Adhemerval Zanella ***@***.***>
> CommitDate: Fri Aug 12 09:15:28 2022 -0300
> 
>      linux: Fix sys/mount.h usage with kernel headers
> 
>      Now that kernel exports linux/mount.h and includes it on linux/fs.h,
>      its definitions might clash with glibc exports sys/mount.h.  To avoid
>      the need to rearrange the Linux header to be always after glibc one,
>      the glibc sys/mount.h is changed to:
> 
>        1. Undefine the macros also used as enum constants.  This covers prior
>           inclusion of <linux/mount.h> (for instance MS_RDONLY).
> 
>        2. Include <linux/mount.h> based on the usual __has_include check
>           (needs to use __has_include ("linux/mount.h") to paper over GCC
>           bugs.
> 
>        3. Define enum fsconfig_command only if FSOPEN_CLOEXEC is not defined.
>           (FSOPEN_CLOEXEC should be a very close proxy.)
> 
>        4. Define struct mount_attr if MOUNT_ATTR_SIZE_VER0 is not defined.
>           (Added in the same commit on the Linux side.)
> 
>      This patch also adds some tests to check if including linux/fs.h and
>      linux/mount.h after and before sys/mount.h does work.
> 
>      Checked on x86_64-linux-gnu.
> 
>      Reviewed-by: Florian Weimer ***@***.***>
> ```
> 
> starting with glibc 2.37...
> 
I installed glibc-2.37-release to a clone of my disk, recompiled gcc, 
and still got the same error.

Again: The error shows not up with util-linux-2.38.1.

Same error with qemu-8.0.0 but not with qemu-7.2.1.
-- 
Regards Klaus

**Comment by karelzak** [comment]
>       3. Define enum fsconfig_command only if FSOPEN_CLOEXEC is not defined.
>          (FSOPEN_CLOEXEC should be a very close proxy.)
> 
>       4. Define struct mount_attr if MOUNT_ATTR_SIZE_VER0 is not defined.
>          (Added in the same commit on the Linux side.)

We can do the same in our fallbacks. See commit 26788963f5dd9bec9a1bdda98896791c61dfad92

@kladitt please, try to rebuild the current git tree on your system. I guess the change will help.

**Comment by thesamesam** [comment]
>I installed glibc-2.37-release to a clone of my disk, recompiled gcc, and still got the same error.

What does glibc-2.37-release mean here? Do you mean the 2.37 release or do you mean the tip of the `releases/glibc-2.37` branch which has this backported _after_ the release?

**Comment by kladitt** [comment]
On 30.04.23 08:53, Sam James wrote:
>> I installed glibc-2.37-release to a clone of my disk, recompiled gcc, and still got the same error.
> 
> What does glibc-2.37-release mean here? Do you mean the 2.37 release or do you mean the tip of the `releases/glibc-2.37` branch which has this backported _after_ the release?
> 

Sam, I do not know what you mean.

For me the term release means and is:

https://ftp.gnu.org/pub/pub/gnu/glibc/glibc-2.37.tar.xz
which is of date 2023-02-01 20:14 	

I do not know of a branch which has this backported (what is this?) 
_after_ the release?

I tried both, with the release and with current

git clone git://sourceware.org/git/glibc.git
(latest commit to from Sat Apr 29 16:12:22 2023 +0300)

and got the same error with both.

-- 
Regeards Klaus

**Comment by kladitt** [comment]
Karel,
good news, current git already installed.
Not direct comparable to before, new kernel, new gcc(13.1.1), glibc-git,  but util-linux works so far here now.
Thank you very much.

**Comment by thesamesam** [comment]
The 'release' branch is a branch based on a tag but has commits backported to it after. But if master failed for you too, that's irrelevant.

**Comment by vnmabus** [comment]
> I installed glibc-2.37-release to a clone of my disk, recompiled gcc, and still got the same error. Again: The error shows not up with util-linux-2.38.1. Same error with qemu-8.0.0 but not with qemu-7.2.1. -- Regards Klaus

I hit the same problem with older GCC. I finally was able to trace it back to this GCC bug: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=91085 . It seems that the `fixincludes` utility in GCC changes the Glibc `mount.h`, replacing inside `__has_include` the string `"linux"` with `"__linux__"`. This causes Glibc `mount.h` not to import the Linux `mount.h` even if present, and thus if the Glibc's  `mount.h` is included before the Linux's one, the compilation fails.

There are several solution to this problem:
- Upgrading GCC to a later version.
- Patching GCC with the patches in https://gcc.gnu.org/git/gitweb.cgi?p=gcc.git;h=6bf383c37e6131a8e247e8a0997d55d65c830b6d .
- Changing the order of the includes so that the Linux's header is always included before the Glibc's header.
